# Vehicle Maintenance Prediction Project



In [22]:
# Import essential libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix
from sklearn.linear_model import LogisticRegression
from sklearn.preprocessing import StandardScaler
from sklearn.ensemble import RandomForestClassifier
import xgboost as xgb
from sklearn.neighbors import KNeighborsClassifier

# Load the dataset
data = pd.read_csv('./data_file/vehicle_maintenance_data.csv')  
data.head()


Vehicle_Model  Mileage Maintenance_History  Reported_Issues  Vehicle_Age  \
0         Truck    58765                Good                0            4   
1           Van    60353             Average                1            7   
2           Bus    68072                Poor                0            2   
3           Bus    60849             Average                4            5   
4           Bus    45742                Poor                5            1   

  Fuel_Type Transmission_Type  Engine_Size  Odometer_Reading  \
0  Electric         Automatic         2000             28524   
1  Electric         Automatic         2500            133630   
2  Electric         Automatic         1500             34022   
3    Petrol         Automatic         2500             81636   
4    Petrol            Manual         2000             97162   

  Last_Service_Date Warranty_Expiry_Date Owner_Type  Insurance_Premium  \
0        2023-11-23           2025-06-24     Second              20782   
1        2023-09-21           2025-06-04     Second              23489   
2        2023-06-27           2025-04-27      First              17979   
3        2023-08-24           2025-11-05     Second               6220   
4        2023-05-25           2025-09-14      Third              16446   

   Service_History  Accident_History  Fuel_Efficiency Tire_Condition  \
0                6                 3        13.622204            New   
1                7                 0        13.625307            New   
2                7                 0        14.306302            New   
3                7                 3        18.709467            New   
4                6                 2        16.977482           Good   

  Brake_Condition Battery_Status  Need_Maintenance  
0             New           Weak                 1  
1             New           Weak                 1  
2            Good           Weak                 1  
3        Worn Out            New                 1  
4            Good           Weak                 1

In [3]:
# Check for missing values
print("Missing values per column:\n", data.isnull().sum())

# Drop rows or handle missing values as needed
data = data.dropna()  # Example: Dropping missing rows

# Display basic statistics
data.describe()

Missing values per column:
 Vehicle_Model           0
Mileage                 0
Maintenance_History     0
Reported_Issues         0
Vehicle_Age             0
Fuel_Type               0
Transmission_Type       0
Engine_Size             0
Odometer_Reading        0
Last_Service_Date       0
Warranty_Expiry_Date    0
Owner_Type              0
Insurance_Premium       0
Service_History         0
Accident_History        0
Fuel_Efficiency         0
Tire_Condition          0
Brake_Condition         0
Battery_Status          0
Need_Maintenance        0
dtype: int64


Mileage  Reported_Issues   Vehicle_Age   Engine_Size  \
count  50000.000000     50000.000000  50000.000000  50000.000000   
mean   54931.232680         2.497420      5.492260   1556.292000   
std    14401.912925         1.708781      2.875682    627.677218   
min    30001.000000         0.000000      1.000000    800.000000   
25%    42471.500000         1.000000      3.000000   1000.000000   
50%    54810.000000         2.000000      5.000000   1500.000000   
75%    67391.500000         4.000000      8.000000   2000.000000   
max    80000.000000         5.000000     10.000000   2500.000000   

       Odometer_Reading  Insurance_Premium  Service_History  Accident_History  \
count      50000.000000       50000.000000     50000.000000       50000.00000   
mean       75551.187060       17465.340700         5.515560           1.50156   
std        43088.105658        7223.393401         2.874899           1.11951   
min         1001.000000        5000.000000         1.000000           0.00000   
25%        38009.000000       11189.750000         3.000000           0.00000   
50%        75598.500000       17477.500000         6.000000           2.00000   
75%       112999.500000       23692.000000         8.000000           3.00000   
max       149999.000000       30000.000000        10.000000           3.00000   

       Fuel_Efficiency  Need_Maintenance  
count     50000.000000      50000.000000  
mean         14.990323          0.809960  
std           2.885583          0.392336  
min          10.000098          0.000000  
25%          12.489037          1.000000  
50%          14.986352          1.000000  
75%          17.474676          1.000000  
max          19.999968          1.000000

In [8]:
# Encode categorical features (if any)
data = pd.get_dummies(data, drop_first=True)

In [23]:
# Define features (X) and target (y)
X = data.drop('Need_Maintenance', axis=1)  # Replace with target column
y = data['Need_Maintenance']

# Split data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)


In [20]:
# Model 3: K-Nearest Neighbors Classifier
knn = KNeighborsClassifier(n_neighbors=5)  # Modify n_neighbors as needed
knn.fit(X_train, y_train)
y_pred_knn = knn.predict(X_test)

# Evaluation metrics
print("K-Nearest Neighbors Accuracy:", accuracy_score(y_test, y_pred_knn))
print("Classification Report:\n", classification_report(y_test, y_pred_knn))
print("Confusion Matrix:\n", confusion_matrix(y_test, y_pred_knn))


K-Nearest Neighbors Accuracy: 0.7837
Classification Report:
               precision    recall  f1-score   support

           0       0.32      0.11      0.17      1915
           1       0.82      0.94      0.88      8085

    accuracy                           0.78     10000
   macro avg       0.57      0.53      0.52     10000
weighted avg       0.72      0.78      0.74     10000

Confusion Matrix:
 [[ 220 1695]
 [ 468 7617]]


### K-Nearest Neighbors Model Summary:

The K-Nearest Neighbors (KNN) classifier achieved an accuracy of 78.37%. However, the precision and recall for predicting the "Need_Maintenance" class (1) were high (0.82 and 0.94), indicating that the model is good at identifying maintenance needs. On the other hand, the precision and recall for class 0 (no maintenance needed) were low (0.32 and 0.11), meaning it struggles to correctly classify instances where no maintenance is required. This could suggest class imbalance, with more focus on correctly predicting class 1.

### Evaluation:
- **Accuracy**: 78.37%
- **Precision**: 0.82 (class 1), 0.32 (class 0)
- **Recall**: 0.94 (class 1), 0.11 (class 0)
- **F1-Score**: 0.88 (class 1), 0.17 (class 0)



In [16]:
#Model 2: Logistic Regression Model

feature_list = ['Vehicle_Model', 'Mileage', 'Maintenance_History', 'Reported_Issues', 'Vehicle_Age', 
                'Fuel_Type', 'Transmission_Type', 'Engine_Size', 'Odometer_Reading', 'Last_Service_Date', 
                'Warranty_Expiry_Date', 'Owner_Type', 'Insurance_Premium', 'Service_History', 
                'Accident_History', 'Fuel_Efficiency', 'Tire_Condition', 'Brake_Condition', 'Battery_Status']

X = data[feature_list]  # Features
y = data['Need_Maintenance']  # Target variable

# Data preprocessing
X = pd.get_dummies(X, drop_first=True)  # Convert categorical features into dummy variables
scaler = StandardScaler()
X = scaler.fit_transform(X)  # Standardize the features

# Split data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Logistic Regression Model
log_reg = LogisticRegression()
log_reg.fit(X_train, y_train)

# Predictions
y_pred = log_reg.predict(X_test)

# Evaluation
print("Accuracy:", accuracy_score(y_test, y_pred))
print("Classification Report:\n", classification_report(y_test, y_pred))
print("Confusion Matrix:\n", confusion_matrix(y_test, y_pred))



Accuracy: 0.9441
Classification Report:
               precision    recall  f1-score   support

           0       0.86      0.84      0.85      1915
           1       0.96      0.97      0.97      8085

    accuracy                           0.94     10000
   macro avg       0.91      0.90      0.91     10000
weighted avg       0.94      0.94      0.94     10000

Confusion Matrix:
 [[1608  307]
 [ 252 7833]]


### Logistic Regression Model Summary:

The Logistic Regression model performed well with an accuracy of 94.41%. The model displayed high precision and recall for class 1 (maintenance needed), with values of 0.96 and 0.97, respectively. For class 0 (no maintenance), the precision and recall were 0.86 and 0.84. Overall, the model showed good balance between correctly identifying both classes, particularly excelling at predicting maintenance needs. 

### Evaluation:
- **Accuracy**: 94.41%
- **Precision**: 0.96 (class 1), 0.86 (class 0)
- **Recall**: 0.97 (class 1), 0.84 (class 0)
- **F1-Score**: 0.97 (class 1), 0.85 (class 0)

In [17]:
#Model 3: Random Forest
rf = RandomForestClassifier(random_state=42)

# Fit the model to the training data
rf.fit(X_train, y_train)

# Predict on the test data
y_pred_rf = rf.predict(X_test)

# Evaluate the model
print("Random Forest Accuracy:", accuracy_score(y_test, y_pred_rf))
print("Classification Report:\n", classification_report(y_test, y_pred_rf))
print("Confusion Matrix:\n", confusion_matrix(y_test, y_pred_rf))

Random Forest Accuracy: 1.0
Classification Report:
               precision    recall  f1-score   support

           0       1.00      1.00      1.00      1915
           1       1.00      1.00      1.00      8085

    accuracy                           1.00     10000
   macro avg       1.00      1.00      1.00     10000
weighted avg       1.00      1.00      1.00     10000

Confusion Matrix:
 [[1915    0]
 [   0 8085]]


### Random Forest Model Summary:

The Random Forest model achieved perfect accuracy of 100%, with both precision and recall for both classes (0: no maintenance, 1: maintenance needed) equal to 1.0. This indicates that the model was able to perfectly classify all instances, without any misclassification. Both the confusion matrix and classification report show ideal performance, where all predictions were correct.

### Evaluation:
- **Accuracy**: 100%
- **Precision**: 1.00 (both classes)
- **Recall**: 1.00 (both classes)
- **F1-Score**: 1.00 (both classes)

In [19]:
#Model 4: XGBoost

model = xgb.XGBClassifier(use_label_encoder=False, eval_metric='mlogloss')

# Train the model
model.fit(X_train, y_train)

# Predict on the test set
y_pred = model.predict(X_test)

# Evaluation
print("XGBoost Accuracy:", accuracy_score(y_test, y_pred))
print("Classification Report:\n", classification_report(y_test, y_pred))
print("Confusion Matrix:\n", confusion_matrix(y_test, y_pred))

/opt/anaconda3/lib/python3.12/site-packages/xgboost/core.py:158: UserWarning: [15:55:37] WARNING: /Users/runner/work/xgboost/xgboost/src/learner.cc:740: 
Parameters: { "use_label_encoder" } are not used.

  warnings.warn(smsg, UserWarning)


XGBoost Accuracy: 1.0
Classification Report:
               precision    recall  f1-score   support

           0       1.00      1.00      1.00      1915
           1       1.00      1.00      1.00      8085

    accuracy                           1.00     10000
   macro avg       1.00      1.00      1.00     10000
weighted avg       1.00      1.00      1.00     10000

Confusion Matrix:
 [[1915    0]
 [   0 8085]]


### XGBoost Model Summary:

The XGBoost model achieved a perfect accuracy of 100%, with both precision and recall for each class (0: no maintenance, 1: maintenance needed) equal to 1.0. The classification report and confusion matrix indicate flawless performance, where all predictions were correct without any misclassifications.

### Evaluation:
- **Accuracy**: 100%
- **Precision**: 1.00 (both classes)
- **Recall**: 1.00 (both classes)
- **F1-Score**: 1.00 (both classes)